### PR TITLE
feat: Introduce protocol service implementation

### PR DESCRIPTION
**Pull request by soluchok**
* added `dispatcher.Outbound`
* tests improvements
* added *decorator.Thread to `Proposal` model
* `process` implementation

Issue: https://github.com/hyperledger/aries-framework-go/issues/464
Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

File: pkg/didcomm/protocol/introduce/states.go [hunk 1]
```diff
@@ -53,11 +65,11 @@ func (s *noOp) CanTransitionTo(_ state) bool {
 	return false
 }
 
-func (s *noOp) ExecuteInbound(_ *metaData) (state, error) {
+func (s *noOp) ExecuteInbound(p provider, _ *metaData) (state, error) {
 	return nil, errors.New("cannot execute no-op")
 }
 
-func (s *noOp) ExecuteOutbound(_ *metaData, _ *service.Destination) (state, error) {
+func (s *noOp) ExecuteOutbound(p provider, _ *metaData, _ *service.Destination) (state, error) {
```
Comment (talwinder50) [inline pkg/didcomm/protocol/introduce/states.go]: Why are we passing provider in execute in bound and outbound ?

File: pkg/didcomm/protocol/introduce/service.go [hunk 2]
```diff
@@ -68,16 +69,20 @@ type Service struct {
 	service.Message
 	store     storage.Store
 	callbacks chan callback
+	sender    *providerImpl
 }
 
 // New returns introduce service
-func New(ctx context.Context, s storage.Provider) (*Service, error) {
+func New(ctx context.Context, s storage.Provider, d dispatcher.Outbound) (*Service, error) {
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/service.go]: We can use a provider similar to [DIDExchange Service](https://github.com/hyperledger/aries-framework-go/blob/master/pkg/didcomm/protocol/didexchange/service.go#L63).

File: pkg/didcomm/protocol/introduce/states.go [hunk 3]
```diff
@@ -29,6 +33,14 @@ const (
 	stateNameWaiting  = "waiting"
 )
 
+type provider interface {
+	dispatcher.Outbound
+}
+
+type providerImpl struct {
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/states.go]: All the dependencies should come from aries.Context(). We shouldn't implement these in protocols.

Comment (troyronda) [reply pkg/didcomm/protocol/introduce/states.go]: (more specifically, the protocol should define an interface of the subset of Context that it needs).

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/states.go]: ok, I will do that more clear :)

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/states.go]: Done!

File: pkg/didcomm/protocol/introduce/service.go (hunk 2 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: Done!

File: pkg/didcomm/protocol/introduce/states.go (hunk 1 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/states.go]: the provider was equal to context (did-exchange), I have renamed it now (provider to internalContext)

File: pkg/didcomm/protocol/introduce/service.go [hunk 4]
```diff
@@ -68,16 +69,26 @@ type Service struct {
...
+// provider contains dependencies for the DID exchange protocol and is typically created by using aries.Context()
+type provider interface {
+	OutboundDispatcher() dispatcher.Outbound
+	StorageProvider() storage.Provider
 }
 
 // New returns introduce service
-func New(ctx context.Context, s storage.Provider) (*Service, error) {
-	store, err := s.OpenStore(Introduce)
+func New(ctx context.Context, p provider) (*Service, error) {
```
Comment (troyronda) [inline pkg/didcomm/protocol/introduce/service.go]: Std lib context.Context should be provided to a request not a constructor.

File: pkg/didcomm/protocol/introduce/service.go [hunk 5]
```diff
@@ -421,13 +440,13 @@ func (s *Service) handle(msg *metaData, dest *service.Destination) error {
 		)
 
 		if dest != nil {
-			followup, err = next.ExecuteOutbound(msg, dest)
+			followup, err = next.ExecuteOutbound(s.ctx, msg, dest)
```
Comment (troyronda) [inline pkg/didcomm/protocol/introduce/service.go]: Context is supposed to be defined for a particular request/transaction - not as a member variable used across transactions. See prior comment.

File: pkg/didcomm/protocol/introduce/service.go (hunk 4 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: I have created it to stop the listener 
https://github.com/hyperledger/aries-framework-go/pull/575/files#diff-43cf3af5d3c6b9acebc5c635ba4d0f7dR97

File: pkg/didcomm/protocol/introduce/service.go (hunk 5 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: it is internal context

File: pkg/didcomm/protocol/introduce/service.go (hunk 4 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: usage here https://github.com/hyperledger/aries-framework-go/blob/master/pkg/didcomm/protocol/introduce/service.go#L110

File: pkg/didcomm/protocol/introduce/service.go (hunk 5 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: the same logic in did-exchange

File: pkg/didcomm/protocol/introduce/service.go (hunk 4 above)
Comment (troyronda) [reply pkg/didcomm/protocol/introduce/service.go]: This seems to be an usual usage of context. Recommend a Stop method instead.

File: pkg/didcomm/protocol/introduce/service.go (hunk 5 above)
Comment (troyronda) [reply pkg/didcomm/protocol/introduce/service.go]: internal context is ok.

File: pkg/didcomm/protocol/introduce/service.go (hunk 4 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: Done!

File: pkg/didcomm/protocol/introduce/service.go [hunk 6]
```diff
@@ -107,16 +134,22 @@ func (s *Service) startInternalListener(ctx context.Context) {
 			if err := s.processFailure(msg.id, msg.err); err != nil {
 				logger.Errorf("process callback : %s", err)
 			}
-		case <-ctx.Done():
+		case <-s.stop:
```
Comment (troyronda) [inline pkg/didcomm/protocol/introduce/service.go]: the log message is out of date.

File: pkg/didcomm/protocol/introduce/service.go [hunk 7]
```diff
@@ -68,28 +69,54 @@ type Service struct {
 	service.Message
 	store     storage.Store
 	callbacks chan callback
+	ctx       internalContext
+	stop      chan struct{}
+	mu        sync.Mutex
```
Comment (troyronda) [inline pkg/didcomm/protocol/introduce/service.go]: i would name the mutex similar to the var it is protecting.

File: pkg/didcomm/protocol/introduce/service.go [hunk 8]
```diff
@@ -130,7 +163,8 @@ func (s *Service) processFailure(id string, pErr error) error {
 }
 
 // abandon updates the state to abandoned and trigger failure event.
-func (s *Service) abandon(thID string, msg *service.DIDCommMsg, _ error) error {
+func (s *Service) abandon(thID string, msg *service.DIDCommMsg, pErr error) error {
+	logger.Infof("go to abandon: %v", pErr)
```
Comment (troyronda) [inline pkg/didcomm/protocol/introduce/service.go]: i assume you are cleaning up log levels in a future PR?

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: removed!

File: pkg/didcomm/protocol/introduce/service.go (hunk 7 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: changed to `closedMutex`

File: pkg/didcomm/protocol/introduce/service.go (hunk 8 above)
Comment (troyronda) [reply pkg/didcomm/protocol/introduce/service.go]: It was a general comment :). There are several logger lines that should not "INFO" level.

**Pull request merged**